process/process.go

```diff
@@ -707,56 +717,58 @@ func (p *Process) insertPropose(propose Propose) bool {
 		return false
 	}
 
+	// We have caught a Process attempting to broadcast two different Proposes at
+	// the same Height and Round. Even though we only explicitly check the Round,
+	// we know that the Proposes will have the same Height, because we only keep
+	// message logs for message with the same Height as the current Height of the
+	// Process.
+	if existingPropose, ok := p.ProposeLogs[propose.Round]; ok {
+		if !propose.Equal(&existingPropose) {
+			if p.catcher != nil {
+				p.catcher.CatchDoublePropose(propose, existingPropose)
+			}
+		}
+		return false
+	}
+
 	if p.scheduler != nil {
 		proposer := p.scheduler.Schedule(propose.Height, propose.Round)
 		if !proposer.Equal(&propose.From) {
-			// We have caught a Process attempting to broadcast a propose
-			// when it was not the scheduled proposer for that height and round.
-			// This is caught as an out of turn propose
+			// We have caught a Process attempting to broadcast a propose when it was
+			// not the scheduled proposer for that height and round. This is caught
+			// as an out of turn propose
 			if p.catcher != nil {
 				p.catcher.CatchOutOfTurnPropose(propose)
 			}
 			return false
 		}
-
-		existingPropose, ok := p.ProposeLogs[propose.Round]
-		if ok {
-			// We have caught a Process attempting to broadcast two different
-			// Proposes at the same Height and Round. Even though we only
-			// explicitly check the Round, we know that the Proposes will have the
-			// same Height, because we only keep message logs for message with the
-			// same Height as the current Height of the Process.
-			if !propose.Equal(&existingPropose) {
-				if p.catcher != nil {
-					p.catcher.CatchDoublePropose(propose, existingPropose)
-				}
-			}
-			return false
-		}
 	}
 
-	// By never inserting a Propose that is not valid, we can avoid the validity
-	// checks elsewhere in the Process.
+	// We discard a nil value proposal. If a validator implementation is provided
+	// we check and store the proposal's validity. In the case of an invalid
+	// proposal, we broadcast a nil prevote, and avoid adding this message to the
+	// trace logs as it is an invalid proposal. We return true as we have in fact
+	// inserted the propose message to our propose logs, while explicitly marking
+	// it as invalid.
 	if propose.Value == NilValue || (p.validator != nil && !p.validator.Valid(propose.Value)) {
 		if p.broadcaster != nil {
 			p.broadcaster.BroadcastPrevote(Prevote{
 				Height: p.CurrentHeight,
-				Round:  p.CurrentRound,
+				Round:  propose.Round,
 				Value:  NilValue,
 				From:   p.whoami,
 			})
-			p.stepToPrevoting()
 		}
```

Comment:
We should not do this broadcast, because it allows an attacker to force everyone to skip to an arbitrary future round. It is also a case that gets handled properly by `tryPrevoteUponPropose` (assuming that we insert invalid proposes, and mark them invalid, which we do).